types/src/state_store/state_key.rs

```diff
@@ -2,22 +2,31 @@
 // SPDX-License-Identifier: Apache-2.0
 
 use crate::{access_path::AccessPath, state_store::table::TableHandle};
+use anyhow::Result;
 use aptos_crypto::{
     hash::{CryptoHash, CryptoHasher},
     HashValue,
 };
 use aptos_crypto_derive::CryptoHasher;
 use num_derive::{FromPrimitive, ToPrimitive};
 use num_traits::FromPrimitive;
-use serde::{Deserialize, Serialize};
-use std::convert::TryInto;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use std::{convert::TryInto, ops::Deref};
 use thiserror::Error;
 
+#[derive(Clone, Debug, CryptoHasher, Eq, PartialEq, Ord, PartialOrd, Hash)]
```

Comment:
Should it derive it if it implements CryptoHasher explicitly (as it should)?

Also, I was always confused of the fact that it also derives Hash, and the data-structures can be using that (e.g. the non-cryptographic hash), which could also bypass the current optimization if this isn't mistaken?

Reply:
The line 139 requires this derive (although I don't think it is used).

This PR is not trying to solve the "Dashmap compute (non-crypto) hash twice" issue. I need to double check whether caching it is a good idea or we should go with a different approach. (e.g. use the first byte of the crypto hash as the shard num, etc.)

For the reference, based on my benchmark, for StateKey, in average, CryptoHash takes ~1ms, Hash takes ~90ns.

Reply:
I see re:Hash, and indeed makes sense that data-structures would not use cryptographic hash, but we could use a similar approach (and should be significantly simpler to add) later to optimize the normal hash as well, right? Just checking my understanding of what's going on.

re:derive, I think it's confusing to have it there (as a reader like myself would falsely assume it wouldn't also be explicitly implemented), but I see line:139 and what you mean. I think you might be able to invoke the 'define_hasher' macro (but it's not exported, so maybe inside hash.rs) and create a fake StateKeyHasher without deriving. There is some context in the aptos_crypto hash.rs documentation.

Reply:
re hash: right

re derive: done, removed the derive, and added a dummy hasher in hash.rs.

Reply:
Thanks!